core/domain/role_domain_test.py

```diff
@@ -0,0 +1,134 @@
+
+import string
+
+from core.domain import role_domain
+from core.tests import test_utils
+
+
+class RoleDomainUnitTests(test_utils.GenericTestBase):
+    """test the role domain"""
+    HIERARCHY = role_domain.ROLE_HIERARCHY
+    ACTIONS = role_domain.ROLE_ACTIONS
+
+    def test_dicts_have_same_keys(self):
+        """test that ROLE_HIERARCHY and ROLE_ACTIONS have same keys
+        in same order.
+        """
+        self.assertEqual(self.HIERARCHY.keys(), self.ACTIONS.keys())
+
+    def test_dicts_have_list_value(self):
+        """test that ROLE_HIERARCHY and ROLE_ACTIONS, both have list as value
+        to all the keys
+        """
+        for i in self.HIERARCHY:
+            self.assertTrue(isinstance(self.HIERARCHY[i], list))
+
+        for i in self.ACTIONS:
+            self.assertTrue(isinstance(self.ACTIONS[i], list))
+
+    def test_valid_names(self):
+        """strings with UPPERCASE alphabets and underscores are allowed. This
+        test checks whether all entries comply with this rule.
+        """
+
+        def valid_character(character):
+            if (character in string.ascii_uppercase) or (character == '_'):
+                return True
+            return False
+
+        for role_name in self.HIERARCHY:
+            for i in role_name:
+                self.assertTrue(valid_character(i))
+
+            for action_name in self.HIERARCHY[role_name]:
+                for i in action_name:
+                    self.assertTrue(valid_character(i))
+
+    def test_every_dict_entry_is_string(self):
+        """test that all keys and values(elements in lists) in ROLE_HIERARCHY
+        and ROLE_ACTIONS are string.
+        """
+
+        # checking the keys in ROLE_HIERARCHY.
+        for i in self.HIERARCHY:
+            self.assertTrue(isinstance(i, str))
+
+            # checking the values in list corresponding to key.
+            for j in self.HIERARCHY[i]:
+                self.assertTrue(isinstance(j, str))
+
+        # checking the keys in ACTIONS.
+        for i in self.ACTIONS:
+            self.assertTrue(isinstance(i, str))
+
+            # checking the values in list corresponding to key.
+            for j in self.ACTIONS[i]:
+                self.assertTrue(isinstance(j, str))
+
+    def test_valid_parents(self):
+        """test that all the roles present in value list for any key in
+        ROLE_HIERARCHY are valid(i.e there exists a key with that name)
+        """
+        valid_roles = self.HIERARCHY.keys()
+
+        for i in self.HIERARCHY:
+            for j in self.HIERARCHY[i]:
+                self.assertTrue(j in valid_roles)
+
+    def test_no_cycles(self):
+        """Visits each role and checks that there is no cycle from that
+        role"""
+        visited = set()
+
+        def check_cycle(source, parents):
+            """checks if there is a cycle starting from source
+            args : source -> the starting role to check
+                   parents -> list of neighbours
```

Comment:
But `parents` isn't always going to be a list of neighbours of `source`. Maybe call it `roles` and describe it as "list(str). A list of roles to start the check from."

Also the overall docstring should be: "Check that `source` is not reachable from any of the given roles."

Finally, please format docstrings according to the standards in #2394.